presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Integer>
+    {
+        private static final String INTERSECT_MARKER = "intersect_marker";
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+
+        private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitIntersect(IntersectNode node, RewriteContext<Integer> rewriteContext)
+        {
+            List<PlanNode> sources = node.getSources();
+            int numberOfSources = sources.size();
+
+            ImmutableList.Builder<PlanNode> newProjectNodes = ImmutableList.builder();
+            ImmutableListMultimap.Builder<Symbol, Symbol> symbolMappings = ImmutableListMultimap.builder();
+            List<Symbol> unionOutputSymbols = null;
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregationsBuilder = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Expression> outputProjections = ImmutableMap.builder();
+
+            for (int i = 0; i < numberOfSources; i++) {
+                ProjectNode newProjectNode = createProjectNode(node, sources, i);
+                newProjectNodes.add(newProjectNode);
+
+                List<Symbol> projectOutputSymbols = newProjectNode.getOutputSymbols();
+                if (unionOutputSymbols == null) {
+                    // Use the first Relation to derive output symbol names
+                    unionOutputSymbols = getUnionOutputSymbolsFrom(projectOutputSymbols);
+                }
+
+                checkArgument(projectOutputSymbols.size() == unionOutputSymbols.size(),
+                        "Intersect output should have %s symbols but has %s symbols",
+                        unionOutputSymbols.size(),
+                        projectOutputSymbols.size());
+
+                int unionFieldId = 0;
+                for (Symbol symbol : projectOutputSymbols) {
+                    symbolMappings.put(unionOutputSymbols.get(unionFieldId), symbol);
+                    unionFieldId++;
+                }
+            }
+
+            UnionNode unionNode = new UnionNode(node.getId(), newProjectNodes.build(), symbolMappings.build(), ImmutableList.copyOf(symbolMappings.build().keySet()));
+            AggregationNode aggregationNode = createAggregationNodeForIntersect(node, unionOutputSymbols, unionNode, aggregationsBuilder, outputProjections);
+            FilterNode filterNode = createFilterNodeForIntersect(aggregationNode, aggregationsBuilder.build());
+
+            return new ProjectNode(idAllocator.getNextId(), filterNode, outputProjections.build());
+        }
+
+        private List<Symbol> getUnionOutputSymbolsFrom(List<Symbol> sourceSymbols)
```

Comment:
it has nothing to with 'UNION'. this method just does kind of a deep-copy of symols. Please rename it.